File: Example Code1/RadioHub/ServiceAccessPoints/DeviceManagement/DeviceManagement.cpp
/**
 * @file    DeviceManagement.cpp
 *
 * @brief   Implementation of class DeviceManagement
 *
 * @note    This example code is free software: you can redistribute it and/or modify it.
 *
 *          This program is provided by IMST GmbH on an "AS IS" basis without
 *          any warranties in the hope that it will be useful.
 */

#include "DeviceManagement.h"
#include <QDateTime>

//<! map with status code strings
const QMap < uint8_t, QString > DeviceManagement::_StatusCodes =
{
    { Ok,                       "ok" },
    { Error,                    "error" },
    { CommandNotSupported,      "command not supported" },
    { WrongParameter,           "wrong parameter" },
    { WrongApplicationMode,     "wrong application mode" },
    { Reserved,                 "reserved" },
    { ApplicationBusy,          "application busy" },
    { WrongMessageLength,       "wrong message length" },
    { NVM_WriteError,           "NVM write error" },
    { NVM_ReadError,            "NVM read error" },
    { CommandRejected,          "command rejected" }
};

//<! map with module type strings
const QMap < uint8_t, QString > DeviceManagement::_ModuleTypes =
{
    { 103,  "iC281A" },
    { 104,  "iM284A-XL" },
    { 163,  "iM881A-XL" }
};

//<! map with event names for HCI messages
const QMap < uint8_t , QString > DeviceManagement::_EventNames =
{
    { Ping_Rsp,                 "ping device response" },
    { GetDeviceInfo_Rsp,        "get device info response" },
    { GetFirmwareVersion_Rsp,   "get firmware version response" },
    { GetDateTime_Rsp,          "get date time response" },
    { SetDateTime_Rsp,          "set date time response" },
    { RestartDevice_Rsp,        "restart device response" }
};

//<! map with message handlers for HCI messages
const QMap < uint8_t , DeviceManagement::Handler > DeviceManagement::_Handlers =
{
    { Ping_Rsp,                 &DeviceManagement::OnDefaultResponse },
    { GetDeviceInfo_Rsp,        &DeviceManagement::OnGetDeviceInfoResponse },
    { GetFirmwareVersion_Rsp,   &DeviceManagement::OnGetFirmwareVersionResponse },
    { GetDateTime_Rsp,          &DeviceManagement::OnGetDateTimeResponse },
    { SetDateTime_Rsp,          &DeviceManagement::OnDefaultResponse },
    { RestartDevice_Rsp,        &DeviceManagement::OnDefaultResponse }
};

/**
 * @brief   class constructor
 *
 * @param   port        a serial port
 */
DeviceManagement::DeviceManagement( QSerialPort& port )
                : ServiceAccessPoint( DeviceManagement_Sap, port )
{
}

/**
 * @brief   send "ping request"
 */
bool
DeviceManagement::OnPingDevice()
{
    return SendMessage( Ping_Req );
}

/**
 * @brief   send "get device information request"
 */
bool
DeviceManagement::OnGetDeviceInfo()
{
    return SendMessage( GetDeviceInfo_Req );
}

/**
 * @brief   send "get firmware version request"
 */
bool
DeviceManagement::OnGetFirmwareVersion()
{
    return SendMessage( GetFirmwareVersion_Req );
}

/**
 * @brief   send "get date time request"
 */
bool
DeviceManagement::OnGetDateTime()
{
    return SendMessage( GetDateTime_Req );
}

/**
 * @brief   send "set date time request"
 */
bool
DeviceManagement::OnSetDateTime()
{
    SerialMessage msg( DeviceManagement_Sap, SetDateTime_Req );

    uint32_t  secondsSincePeriod = QDateTime::currentSecsSinceEpoch();

    msg.Append( secondsSincePeriod );

    return SendMessage( msg );
}

/**
 * @brief   send "restart device request"
 */
bool
DeviceManagement::OnRestartDevice()
{
    return SendMessage( RestartDevice_Req );
}

/**
 * @brief   find message handler and decode convert message into json format
 *
 * @param   serialMsg       incoming HCI message
 *
 * @param   result          decoded data in Json format
 *
 * @return  true/false
 */
bool
DeviceManagement::OnDecodeMessage( const SerialMessage& serialMsg, QJsonObject& result )
{
    uint8_t msgID = serialMsg.GetMsgID();

    // find handler by message ID
    Handler handler = _Handlers.value( msgID, nullptr );
    if ( handler != nullptr )
    {
        // get handler event name
        result[ "Event" ] = QString( _EventNames.value( msgID, "unknown handler" ) );

        // call message handler
        return (this->*handler)( serialMsg, result );
    }
    // nothing decoced
    return false;
}

/**
 * @brief   decode default response, HCI status only
 *
 * @param   serialMsg       incoming HCI message
 *
 * @param   result          decoded data in Json format
 *
 * @return  true/false
 */
bool
DeviceManagement::OnDefaultResponse( const SerialMessage& serialMsg, QJsonObject& result )
{
    // check minimum payload length
    if ( serialMsg.GetPayloadLength() >= 1 )
    {
        result[ "Status" ] = _StatusCodes.value( serialMsg.GetResponseStatus(), "error" );

        return true;
    }
    return false;
}

/**
 * @brief   decode device information response
 *
 * @param   serialMsg       incoming HCI message
 *
 * @param   result          decoded data in Json format
 *
 * @return  true/false
 */
bool
DeviceManagement::OnGetDeviceInfoResponse( const SerialMessage& serialMsg, QJsonObject& result )
{
    // check minimum payload length
    if ( serialMsg.GetPayloadLength() >= ( 1 + 1 + 4 ) )
    {
        uint8_t status = serialMsg.GetResponseStatus();

        result[ "Status" ] = _StatusCodes.value( status, "error" );

        if ( status == Ok )
        {
            int index               =   SerialMessage::Status_Index + 1;
            uint8_t moduleType      =   serialMsg.GetU8( index++ );
            uint32_t moduleID       =   serialMsg.GetU32( index ); index += 4;

            QJsonObject info;
            info[ "Module Type" ]   =   _ModuleTypes.value( moduleType, "unknown module type" );
            info[ "Module ID" ]     =   QString::number( moduleID, 16 );

            result[ "Device Info" ] =   info;
        }
        return true;
    }
    return false;
}

/**
 * @brief   decode firmware version response
 *
 * @param   serialMsg       incoming HCI message
 *
 * @param   result          decoded data in Json format
 *
 * @return  true/false
 */
bool
DeviceManagement::OnGetFirmwareVersionResponse( const SerialMessage& serialMsg, QJsonObject& result )
{
    // check minimum payload length
    if ( serialMsg.GetPayloadLength() >= ( 1 + 1 + 1 + 2 + 10 ) )
    {
        uint8_t status = serialMsg.GetResponseStatus();

        result[ "Status" ] = _StatusCodes.value( status, "error" );

        if ( status == Ok )
        {
            int index               =   SerialMessage::Status_Index + 1;
            uint8_t  minorVersion   =   serialMsg.GetU8( index++ );
            uint8_t  majorVersion   =   serialMsg.GetU8( index++ );
            uint16_t  buildCount    =   serialMsg.GetU16( index ); index += 2;
            QByteArray buildDate    =   serialMsg.GetPayload( index, 10 ); index += 10;
            QByteArray name         =   serialMsg.GetPayload( index );

            QJsonObject info;
            info[ "Version" ]       =   QString::number( majorVersion ) + "." + QString::number( minorVersion );
            info[ "Build Count" ]   =   QString::number ( buildCount );
            info[ "Build Date" ]    =   QString( buildDate );
            info[ "Firmware Name" ] =   QString( name );

            result[ "Firmware Info" ] = info;
        }
        return true;
    }
    return false;
}

/**
 * @brief   decode get date time response
 *
 * @param   serialMsg       incoming HCI message
 *
 * @param   result          decoded data in Json format
 *
 * @return  true/false
 */
bool
DeviceManagement::OnGetDateTimeResponse( const SerialMessage& serialMsg, QJsonObject& result )
{
    // check minimum payload length
    if ( serialMsg.GetPayloadLength() >= ( 1 + 4 ) )
    {
        result[ "Status" ]      =   _StatusCodes.value( serialMsg.GetResponseStatus(), "error" );

        uint32_t seconds        =   serialMsg.GetU32( SerialMessage::Status_Index + 1 );

        QJsonObject info;
        info[ "Seconds since epoch" ] =   QString::number( seconds );

        QDateTime dt;
        dt.setSecsSinceEpoch( seconds );
        dt.setTimeSpec( Qt::LocalTime );

        info[ "Date Time" ]     =   dt.toString( "dd-MM-yyyy hh:mm:ss" );

        result[ "Date Time Info" ] =    info;

        return true;
    }
    return false;
}

File: Example Code1/RadioHub/ServiceAccessPoints/DeviceManagement.py
'''
 @file    DeviceManagement.py
 
 @brief   Implementation of DeviceManagement message encoders and decoders
 
 @note    This example code is free software: you can redistribute it and/or modify it.
 
          This program is provided by IMST GmbH on an "AS IS" basis without
          any warranties in the hope that it will be useful.
 '''

from enum import IntEnum

from RadioHub.ServiceAccessPoints.ServiceAccessPoint import ServiceAccessPoint
from RadioHub.Utils.SerialMessage import SerialMessage

from datetime import datetime, timezone
import time

import struct

class DeviceManagement( ServiceAccessPoint ):
    # message identifier of DeviceManegemnt SAP
    class Msg_ID( IntEnum ):
        PingDevice_Req          =   0x01
        PingDevice_Rsp          =   0x02
        GetDeviceInfo_Req       =   0x03
        GetDeviceInfo_Rsp       =   0x04
        GetFirmwareInfo_Req     =   0x05
        GetFirmwareInfo_Rsp     =   0x06
        RestartDevice_Req       =   0x07
        RestartDevice_Rsp       =   0x08
        SetDateTime_Req         =   0x0D
        SetDateTime_Rsp         =   0x0E
        GetDateTime_Req         =   0x0F
        GetDateTime_Rsp         =   0x10

    # status codes of DeviceManagement SAP
    class Status_ID( IntEnum ):
        Ok                      =   0x00
        Error                   =   0x01
        CommandNotSupported     =   0x02
        WrongParameter          =   0x03
        WrongApplicationMode    =   0x04
        Reserved                =   0x05
        ApplicationBusy         =   0x06
        WrongMessageLength      =   0x07
        NVM_WriteError          =   0x08
        NVM_ReadError           =   0x09
        CommandRejected         =   0x0A

    class Module_ID( IntEnum ):
        iC281A                  =   103
        iM284A_XL               =   104
        iM881A_XL               =   163

    # send a simple request
    def defaultRequest( self, msg_id ):
        super().sendRequest( msg_id = msg_id )

    # send encoded date time
    def setDateTimeRequest( self, msg_id ):
        secondsSinceEpoch = int( datetime.now( timezone.utc ).timestamp() )
        payload = struct.pack('<I', secondsSinceEpoch )  
        super().sendRequest( msg_id = msg_id, payload = payload )

    # dictinary with message encoders
    Encoders = {
        "pingDevice"            :  { "ID" : Msg_ID.PingDevice_Req ,         "handler" : defaultRequest },
        "getDeviceInfo"         :  { "ID" : Msg_ID.GetDeviceInfo_Req ,      "handler" : defaultRequest },
        "getFirmwareInfo"       :  { "ID" : Msg_ID.GetFirmwareInfo_Req ,    "handler" : defaultRequest },
        "getDateTime"           :  { "ID" : Msg_ID.GetDateTime_Req ,        "handler" : defaultRequest },
        "setDateTime"           :  { "ID" : Msg_ID.SetDateTime_Req ,        "handler" : setDateTimeRequest },
        "restartDevice"         :  { "ID" : Msg_ID.RestartDevice_Req ,      "handler" : defaultRequest },
    }    

    ModuleTypes = {
        Module_ID.iC281A    :   "iC281A",
        Module_ID.iM284A_XL :   "iM284A-XL",
        Module_ID.iM881A_XL :   "iM881A-XL",
    }

    # default response decoder -> call super
    def decodeDefaultResponse( self, serialMsg : SerialMessage, result : dict ):
        super().decodeDefaultResponse( serialMsg, result = result )

    # decode date and time response
    def decodeDeviceInfoResponse( self, serialMsg : SerialMessage, result : dict ):
        self.decodeDefaultResponse( serialMsg, result )
        index = SerialMessage.Status_Index + 1
        moduleType = serialMsg.getU8( index )
        moduleID = serialMsg.getU32( index + 1 )
        result[ "module type" ] = self.ModuleTypes.get( moduleType, "unknown module type" )
        result[ "module ID" ] = hex( moduleID )

    # decode firmware version response
    def decodeFirmwareVersionResponse( self, serialMsg : SerialMessage, result : dict ):
        self.decodeDefaultResponse( serialMsg, result )
        index = SerialMessage.Status_Index + 1
        minor = serialMsg.getU8( index )
        major = serialMsg.getU8( index + 1 )
        result[ "version" ]         = str( major ) + '.' + str( minor )
        result[ "build count" ]     = str( serialMsg.getU16( index + 2 ) )
        result[ "build date" ]      = serialMsg.getString( index + 4, 10 )
        result[ "firmware name" ]   = serialMsg.getString( index + 14, -1 )

    # decode get date and time response
    def decodeDateTimeResponse( self, serialMsg : SerialMessage, result : dict ):
        self.decodeDefaultResponse( serialMsg, result )
        index = SerialMessage.Status_Index + 1
        dateTime = serialMsg.getDateTime( index )
        result[ "date time" ] =  time.strftime( '%d.%m.%Y %H:%M:%S', dateTime )

    # dictinary with message decoders 
    ResponseDecoders = {
        Msg_ID.PingDevice_Rsp       : { "name" : "pingDevice response",         "handler" : decodeDefaultResponse },
        Msg_ID.GetDeviceInfo_Rsp    : { "name" : "getDeviceInfo response",      "handler" : decodeDeviceInfoResponse },
        Msg_ID.GetFirmwareInfo_Rsp  : { "name" : "getFirmwareVersion response", "handler" : decodeFirmwareVersionResponse },
        Msg_ID.GetDateTime_Rsp      : { "name" : "getDateTime response",        "handler" : decodeDateTimeResponse },
        Msg_ID.SetDateTime_Rsp      : { "name" : "setDateTime response",        "handler" : decodeDefaultResponse },
        Msg_ID.RestartDevice_Rsp    : { "name" : "restartDevice response",      "handler" : decodeDefaultResponse },
    }

    # dictionary with status codes
    StatusCodes = {
        Status_ID.Ok                    :   "ok",
        Status_ID.Error                 :   "error",
        Status_ID.CommandNotSupported   :   "command not supported",
        Status_ID.WrongParameter        :   "wrong parameter",
        Status_ID.WrongApplicationMode  :   "wrong application mode",
        Status_ID.Reserved              :   "reserved",
        Status_ID.ApplicationBusy       :   "application busy",
        Status_ID.WrongMessageLength    :   "wrong message length",
        Status_ID.NVM_WriteError        :   "NVM write error",
        Status_ID.NVM_ReadError         :   "NVM read error",
        Status_ID.CommandRejected       :   "Command rejected",
    }

    def __init__( self, client, serialDevice, sap_id ):
        super().__init__( client            = client,
                          serialDevice      = serialDevice, 
                          sap_id            = sap_id, 
                          responseDecoders  = self.ResponseDecoders, 
                          statusCodes       = self.StatusCodes,
                          encoders          = self.Encoders )
